src/flags.zig

```diff
@@ -302,84 +302,160 @@ fn parse_value(comptime T: type, flag: []const u8, value: [:0]const u8) T {
     comptime unreachable;
 }
 
-pub const ByteSize = struct { bytes: u64 };
 fn parse_value_size(flag: []const u8, value: []const u8) ByteSize {
     assert((flag[0] == '-' and flag[1] == '-') or flag[0] == '<');
 
-    const units = .{
-        .{ &[_][]const u8{ "TiB", "tib" }, 1024 * 1024 * 1024 * 1024 },
-        .{ &[_][]const u8{ "GiB", "gib" }, 1024 * 1024 * 1024 },
-        .{ &[_][]const u8{ "MiB", "mib" }, 1024 * 1024 },
-        .{ &[_][]const u8{ "KiB", "kib" }, 1024 },
-    };
-
-    const unit: struct { suffix: []const u8, scale: u64 } = unit: inline for (units) |unit| {
-        const suffixes = unit[0];
-        const scale = unit[1];
-        for (suffixes) |suffix| {
-            if (std.mem.endsWith(u8, value, suffix)) {
-                break :unit .{ .suffix = suffix, .scale = scale };
-            }
-        }
-    } else break :unit .{ .suffix = "", .scale = 1 };
-
-    assert(std.mem.endsWith(u8, value, unit.suffix));
-    const value_numeric = value[0 .. value.len - unit.suffix.len];
-
-    const amount = std.fmt.parseUnsigned(u64, value_numeric, 10) catch |err| {
+    return ByteSize.parse(value) catch |err| {
         switch (err) {
-            error.Overflow => fatal(
+            error.ParseOverflow => fatal(
                 "{s}: value exceeds 64-bit unsigned integer: '{s}'",
                 .{ flag, value },
             ),
-            error.InvalidCharacter => fatal(
-                "{s}: expected a size, but found '{s}' (invalid digit or suffix)",
+            error.MissingDigits => fatal(
+                "{s}: need digits in size '{s}'",
+                .{ flag, value },
+            ),
+            error.InvalidUnit => fatal(
+                "{s}: invalid unit in size '{s}', (needed KiB, MiB, GiB or TiB)",
+                .{ flag, value },
+            ),
+            error.BytesOverflow => fatal(
+                "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
                 .{ flag, value },
             ),
         }
     };
-
-    const bytes = std.math.mul(u64, amount, unit.scale) catch fatal(
-        "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
-        .{ flag, value },
-    );
-    return ByteSize{ .bytes = bytes };
 }
 
+pub const ByteUnit = enum(u64) {
+    bytes = 1,
+    kib = 1024,
+    mib = 1024 * 1024,
+    gib = 1024 * 1024 * 1024,
+    tib = 1024 * 1024 * 1024 * 1024,
+};
+
+const ByteSizeParseError = error{
+    ParseOverflow,
+    MissingDigits,
+    InvalidUnit,
+    BytesOverflow,
+};
+
+pub const ByteSize = struct {
+    value: u64,
+    unit: ByteUnit = .bytes,
+
+    fn parse(value: []const u8) ByteSizeParseError!ByteSize {
+        assert(value.len != 0);
+
+        const units = .{
+            .{ &[_][]const u8{ "TiB", "tib" }, ByteUnit.tib },
+            .{ &[_][]const u8{ "GiB", "gib" }, ByteUnit.gib },
+            .{ &[_][]const u8{ "MiB", "mib" }, ByteUnit.mib },
+            .{ &[_][]const u8{ "KiB", "kib" }, ByteUnit.kib },
+        };
```

Comment:
Now that we have split the unit into a dedicated `[]const u8` slice, what do you think about using `std.ascii.endsWithIgnoreCase` and eliminating the special cases (i.e. "TiB" and "tib" )? 

I think it's okay if someone inputs "1GIB" or "1Gib" instead of "1GiB."

Reply:
Done.